packages/kit/src/core/sync/write_env.js

```diff
@@ -0,0 +1,183 @@
+import { write_if_changed } from './utils.js';
+import path from 'path';
+import fs from 'fs';
+import { loadEnv } from 'vite';
+
+const autogen_comment = '// this section is auto-generated';
+
+/**
+ * @param {boolean} pub
+ * @param {Record<string, string>} env
+ * @returns {string}
+ */
+function const_declaration_template(pub, env) {
+	return `${autogen_comment}
+
+${Object.entries(env)
+	.map(
+		([k, v]) => `/**
+ * @type {import('$app/env${pub ? '' : '/private'}').${k}}
+ */
+export const ${k} = ${JSON.stringify(v)};
+`
+	)
+	.join('\n')}
+`;
+}
+
+/**
+ * @param {boolean} pub
+ * @param {Record<string, string>} env
+ * @returns {string}
+ */
+function type_declaration_template(pub, env) {
+	return `declare module '$app/env${pub ? '' : '/private'}' {
+  ${Object.keys(env)
+		.map((k) => `export const ${k}: string;`)
+		.join('\n  ')}
+}`;
+}
+
+function runtime_env_template() {
+	return `/** @type {App.RuntimeEnv} */
+export let env = {};
+
+/** @type {(environment: Record<string, string>) => void} */
+export function set_env(environment) {
+  env = environment;
+}`;
+}
+
+/**
+ * Writes the existing environment variables in process.env to
+ * $app/env and $app/env/private
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {string | undefined} mode
+ * The Vite mode.
+ */
+export function write_env(config, mode = undefined) {
+	const pub = write_public_env(config, mode);
+	const priv = write_private_env(config, mode);
+	write_runtime_env(config);
+	write_typedef(config, pub, priv);
+}
+
+/**
+ * Writes the existing environment variables prefixed with config.kit.env.publicPrefix
+ * in process.env to $app/env
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {string | undefined} mode
+ * The Vite mode.
+ */
+function write_public_env(config, mode = undefined) {
+	const pub_out = path.join(config.outDir, 'runtime/app/env.js');
+
+	// public is a little difficult since we append to an
+	// already-existing file
```

Comment:
what if we changed it to `$app/env/public`? it's arguably cleaner anyway

Reply:
I suggested this at one point and Ben wanted to stick with `$app/env` since it's already a thing, but given that we have `/private` and `/runtime` now, I think I prefer `/public`

Reply:
yeah, it's way simpler. i made that change